fl4health/clients/fenda_client.py

```diff
@@ -1,12 +1,226 @@
+        checkpointer: Optional[TorchCheckpointer] = None,
+        temperature: Optional[float] = 0.5,
+        perfcl_loss_weights: Optional[Tuple[float, float]] = (0.0, 0.0),
+        cos_sim_loss_weight: Optional[float] = 0.0,
+        contrastive_loss_weight: Optional[float] = 0.0,
+    ) -> None:
+        super().__init__(
+            data_path=data_path,
+            metrics=metrics,
+            device=device,
+            loss_meter_type=loss_meter_type,
+            metric_meter_type=metric_meter_type,
+            checkpointer=checkpointer,
+        )
+        """This module is used to init fenda client with various auxiliary loss functions.
+        These losses will be activated only when their weights are not 0.0.
+        Args:
+            data_path: Path to the data directory.
+            metrics: List of metrics to be used for evaluation.
+            device: Device to be used for training.
+            loss_meter_type: Type of loss meter to be used.
+            metric_meter_type: Type of metric meter to be used.
+            checkpointer: Checkpointer to be used for checkpointing.
+            temperature: Temperature to be used for contrastive loss.
+            perfcl_loss_weights: Weights to be used for perfcl loss.
+            Each value associate with one of two contrastive losses in perfcl loss.
+            cos_sim_loss_weight: Weight to be used for cosine similarity loss.
+            contrastive_loss_weight: Weight to be used for contrastive loss.
+        """
+        self.perfcl_loss_weights = perfcl_loss_weights
+        self.cos_sim_loss_weight = cos_sim_loss_weight
+        self.contrastive_loss_weight = contrastive_loss_weight
+        self.cos_sim = torch.nn.CosineSimilarity(dim=-1).to(self.device)
+        self.ce_criterion = torch.nn.CrossEntropyLoss().to(self.device)
+        self.temperature = temperature
+
+        # Need to save previous local module, global module and aggregated global module at each communication round
+        # to compute contrastive loss.
+        self.old_local_module: Optional[torch.nn.Module] = None
+        self.old_global_module: Optional[torch.nn.Module] = None
+        self.aggregated_global_module: Optional[torch.nn.Module] = None
+
     def get_parameter_exchanger(self, config: Config) -> ParameterExchanger:
         assert isinstance(self.model, FendaModel)
         return FixedLayerExchanger(self.model.layers_to_exchange())
+
+    def predict(self, input: torch.Tensor) -> Dict[str, torch.Tensor]:
+
+        preds = self.model(input)
+
+        if self.contrastive_loss_weight or self.perfcl_loss_weights:
+            if self.old_local_module is not None:
+                preds["old_local_features"] = self.old_local_module.forward(input).reshape(
+                    len(preds["local_features"]), -1
+                )
+                if self.perfcl_loss_weights:
+                    if self.old_global_module is not None:
```

Comment:
If these modules aren't defined, the PerFCL loss isn't going to be calculable, at least in a correct way right? Rather than doing an if here, where we'll fail somewhere else (or worse get through with a bad calculation), I'd suggest we do
```
assert self.old_global_module is not None and self.aggregated_global_module is not None
```

Reply:
I don't think assertion here would be a wise choice as in the first round of training we won't have old module, but it is fine to compute one round without contrastive loss.

Reply:
Ah good point. Keep it as is then!